### DISCUSSION TITLE
Inject this.socket from outside the Ros class

### DISCUSSION THREAD

**Discussion by eyr1n** [RobotWebTools/roslibjs#798] [Ideas]
Hi,

Although ROSLIBjs is a good client that allows you to communicate with ROS easily, I sometimes feel that rosbridge is lacking in performance.

Therefore, we would like to be able to inject a `socket` member into the `Ros` class from the constructor, etc., so that it can communicate with any bridge.

For example, Foxglove's ros-foxglove-bridge is a very good bridge in the aspect of performance, but if we want to use it in ROSLIBjs, we need to override the members by using inheritance.

The following repository shows an example of ros-foxglove-bridge support by forcibly overriding the `socket` member of the `Ros` class.

https://github.com/eyr1n/roslib-with-foxglove

```ts
export class FoxgloveRos extends Ros {
  declare socket: FoxgloveSocket | null;

  constructor(options: {
    url?: string;
  }) {
    super({
      ...options,
      transportOptions: {
        encoder: (
          message: unknown,
          sendEncodedMessage: (message: unknown) => void,
        ) => sendEncodedMessage(message),
      } as never,
    });
  }

  connect(url: string) {
    if (!this.socket || this.socket.readyState === WebSocket.CLOSED) {
      this.socket = new FoxgloveSocket(url, this);
    }
  }
}
```

I thought it would be a good idea to define the interface and allow the injection of the `socket` member so that such code can be implemented safely.

Thank you.

**Comment by EzraBrooks** [comment]
I would be happy to review a PR that refactors the socket interface to allow more approaches like this one - I think the constructor and connection logic of the Ros class are overly complicated and could be simplified by defining a concrete interface for socket transports like you suggest.

**Comment by EzraBrooks** [comment]
Looks like #1042 will implement this!
